### PR TITLE
Deprecate `.length` in favor of `.size`

### PR DESCRIPTION
**Pull request by bradcray**
In response to users' concerns about having multiple ways to say the same thing, it was proposed and generally agreed-upon in issue #13746 to replace `.length` queries with `.size` for uniformity across types.  This PR implements that by:

* adding deprecation warnings and documentation to Bytes.length (param and non-param versions) (Bytes.chpl)
* adding a deprecation warning to c_array.length (CPtr.chpl)
* adding deprecation warnings to c_string.length (param and non-param versions) and adding a c_string.size param overload (CString.chpl)
* adding a deprecation warning and documentation version of range.length and repurposing range.size to use the old range.length implementation (ChapelRange.chpl)
* adding deprecation warnings and documentation to string.length (param and non-param versions) (String.chpl)
* adding a deprecation warning and documentation to CollectionImpl.length (Collection.chpl)
* adding a `file.size` method and deprecating `file.length()` (IO.chpl)
* changing the field `length` to `size` in LinkedLists, replacing the `.size` method with `.length` and adding deprecation warnings for it (LinkedLists.chpl)
* changing the field `length` to `size` in `reMatch` and adding a `.length` method to generate deprecation warnings for it; note that this also affects the way these records are printed, since they use the default writer which prints the field names (Regexp.chpl)
* updating tons of uses of `.length` in code, documentation, output / debugging output

While here, I also:
* improved a zippered size mismatch error message that used the term "length" for clarity/brevity (ChapelRange.chpl)
* updated an error message about length not being supported on unbounded ranges (ChapelRange.chpl)

TODO:
- [x] check rendered docs
- [x] add deprecation warning tests

[Editorializing: personally, I continue to find `.length` a lot more intuitive in many 1D cases than `.size`, and based on the number of tests that I touched in this PR, I'd guess others feel the same.  If there's a huge user backlash against this deprecation, I'd propose that we enable `.length` for anything 1D (but not permit it for multidimensional things).  This would arguably distinguish it from `.size` in a minor way while keeping the more intuitive term available for people.]

### CODE REVIEW

**Comment by bradcray** [comment]
@e-kayrakli : This is a thankless ask, but would you be willing to review this PR for me?  Your familiarity with strings/bytes (as well as many of the other involved types) is what brought you to mind.  It's technically quite straightforward, but a bit overwhelming in terms of the overall size of the diff (and worth at least some checking to make sure I didn't screw things up).

File: test/deprecated/length.good [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+length.chpl:23: warning: 'string.length' is deprecated - please use 'string.size' instead
+length.chpl:24: warning: 'string.length' is deprecated - please use 'string.size' instead
+length.chpl:25: warning: 'string.length' is deprecated - please use 'string.size' instead
+length.chpl:26: warning: 'c_string.length' is deprecated - please use 'c_string.size' instead
+length.chpl:27: warning: 'c_array.length' is deprecated - please use 'c_array.size' instead
+length.chpl:28: warning: 'range.length' is deprecated - please use 'range.size' instead
+length.chpl:29: warning: 'file.length()' is deprecated - please use 'file.size' instead
+length.chpl:30: warning: 'LinkedList.length' is deprecated - please use 'LinkedList.size' instead
+length.chpl:31: warning: 'reMatch.length' is deprecated - please use 'reMatch.size' instead
+length.chpl:32: warning: 'CollectionImpl.length' is deprecated - please use 'CollectionImpl.size' instead
```
Comment (e-kayrakli) [inline test/deprecated/length.good]: I had issues with too many compiler warnings in single deprecation tests, where some of the warnings were dropped randomly. I'd probably run this test multiple times out of paranoia

File: test/types/range/enum/enumRangeStrided.chpl [hunk 2]
```diff
@@ -13,7 +13,7 @@ proc testRange(r) {
     writeln(c);
   writeln();
 
-  writeln("length = ", r.length);
+  writeln("length = ", r.size);
```
Comment (e-kayrakli) [inline test/types/range/enum/enumRangeStrided.chpl]: Maybe drop this one, too?

File: modules/dists/BlockCycDist.chpl [hunk 3]
```diff
@@ -567,7 +567,7 @@ iter BlockCyclicDom.these(param tag: iterKind) where tag == iterKind.leader {
           temp     = temp.chpl__unTranslate(dimLow);
 
           retblock(j) = (temp.low / dim.stride:idxType)..
-                        #temp.length;
```
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/dists/BlockCycDist.chpl]: You can't expect me to atone for _all_ the mistakes of the past (even if they might have been mine).  :D

File: modules/internal/CPtr.chpl [hunk 4]
```diff
@@ -198,7 +198,12 @@ module CPtr {
       ch <~> new ioLiteral("]");
     }
 
+    /*
+      Deprecated - please use `c_array.size`.
```
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/internal/CPtr.chpl]: I don't think it's a proc though...?  Will that still work?  Or is there another tag to use to hyperlink to a param declaration?  (I was admittedly too lazy to look it up, not being all that invested in documentation that'll be going away soon enough).

File: modules/dists/BlockCycDist.chpl (hunk 3 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/dists/BlockCycDist.chpl]: Fair enough :)

File: modules/packages/Collection.chpl [hunk 5]
```diff
@@ -162,10 +162,10 @@ module Collection {
...
     }
 
-    /*
-      Syntactic sugar for `getSize`.
-    */
+    /* Deprecated - please use :proc:`CollectionImpl.size`. */
     proc length : int {
+      compilerWarning("'CollectionImpl.length' is deprecated - " +
+                      "please use 'CollectionImpl.size' instead");
       return getSize();
```
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/packages/Collection.chpl]: I must admit I don't really "get" the Collection module...

File: modules/dists/dims/BlockDim.chpl [hunk 6]
```diff
@@ -106,7 +106,7 @@ proc BlockDim.init(numLocales: int, boundingBox: range(?),
 
   this.numLocales = numLocales;
   this.bbStart = boundingBox.low;
-  this.bbLength = boundingBox.length;
+  this.bbLength = boundingBox.size;
```
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/dists/dims/BlockDim.chpl]: You almost convinced me to do this, then I thought about the additional testing I'd need to do, so I'm going to not take the option.

File: test/types/range/enum/enumRangeStrided.chpl (hunk 2 above)
Comment (bradcray) [reply test/types/range/enum/enumRangeStrided.chpl]: Oh, good call.

File: modules/internal/CPtr.chpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/internal/CPtr.chpl]: OK, you guilted me into it:  `:var:` works.

File: test/deprecated/length.good (hunk 1 above)
Comment (bradcray) [reply test/deprecated/length.good]: 100 trials pass for me.  What kind of lame warnings were you generating?  :D

Comment (e-kayrakli) [reply test/deprecated/length.good]: Maybe they were too gentle...

**Pull request merged**